torch/distributions/kl.py

```diff
@@ -289,17 +293,46 @@ def _kl_laplace_laplace(p, q):
     return t1 + t2 + t3 - 1
 
 
+@register_kl(LowRankMultivariateNormal, LowRankMultivariateNormal)
+def _kl_lowrankmultivariatenormal_lowrankmultivariatenormal(p, q):
+    if p.event_shape != q.event_shape:
+        raise ValueError("KL-divergence between two Low Rank Multivariate Normals with\
+                          different event shapes cannot be computed")
+
+    term1 = (_batch_lowrank_logdet(q.scale_factor, q.scale_diag, q._capacitance_tril) -
+             _batch_lowrank_logdet(p.scale_factor, p.scale_diag, p._capacitance_tril))
+    term3 = _batch_lowrank_mahalanobis(q.scale_factor, q.scale_diag, q.loc - p.loc,
+                                       q._capacitance_tril)
+    # Expands term2 according to
+    # inv(qcov) @ pcov = [inv(qD) - inv(qD) @ qW @ inv(qC) @ qW.T @ inv(qD)] @ (pW @ pW.T + pD)
+    #                  = [inv(qD) - A.T @ A] @ (pD + pW @ pW.T)
+    qWt_qDinv = q.scale_factor.transpose(-1, -2) / q.scale_diag.unsqueeze(-2)
+    A = _batch_trtrs_lower(qWt_qDinv, q._capacitance_tril)
+    term21 = (p.scale_diag / q.scale_diag).sum(-1)
+    term22 = _batch_trace_XXT(p.scale_factor * q.scale_diag.rsqrt().unsqueeze(-1))
+    term23 = _batch_trace_XXT(A * p.scale_diag.sqrt().unsqueeze(-2))
+    term24 = _batch_trace_XXT(A.matmul(p.scale_factor))
+    term2 = term21 + term22 - term23 - term24
+    return 0.5 * (term1 + term2 + term3 - p.event_shape[0])
+
+
 @register_kl(MultivariateNormal, MultivariateNormal)
 def _kl_multivariatenormal_multivariatenormal(p, q):
     # From https://en.wikipedia.org/wiki/Multivariate_normal_distribution#Kullback%E2%80%93Leibler_divergence
     if p.event_shape != q.event_shape:
         raise ValueError("KL-divergence between two Multivariate Normals with\
                           different event shapes cannot be computed")
 
-    term1 = _batch_diag(q.scale_tril).log().sum(-1) - _batch_diag(p.scale_tril).log().sum(-1)
-    term2 = _batch_trace_XXT(torch.matmul(_batch_inverse(q.scale_tril), p.scale_tril))
-    term3 = _batch_mahalanobis(q.scale_tril, (q.loc - p.loc))
-    return term1 + 0.5 * (term2 + term3 - p.event_shape[0])
+    half_term1 = (_batch_diag(q._unbroadcasted_scale_tril).log().sum(-1) -
```

Comment:
Do I understand correctly that the changes here are simply to support cheaper unbroadcasted computations?

Have you added any new test cases to exercise the broadcasting logic?

Reply:
For your first question, yes.
For the second, I think that the broadcasting logic is implicit from the previous version: `p.scale_tril` and `q.scale_tril` should be broadcasted.

Anyway, I will add a kl test for two distributions with different `batch_shape` (but can be broadcastable).